src/k8s/vmBuilder.js

```diff
@@ -282,3 +282,86 @@ export const getDataVolumeTemplateSpec = (storage, dvSource) => {
     },
   };
 };
+
+export const assignBootOrderIndex = (vm, currDevBootOrder = -1) => {
+  let bootOrder = currDevBootOrder;
+  if (currDevBootOrder !== BOOT_ORDER_FIRST) {
+    bootOrder = getNextAvailableBootOrderIndex(vm);
+  }
+  return bootOrder;
+};
+
+const getNextAvailableBootOrderIndex = vm => {
+  const disks = getDisks(vm);
+  const nics = getInterfaces(vm);
+  let largestIdx = -1;
+
+  disks.forEach(disk => {
+    const bootOrder = get(disk, 'bootOrder', -1);
+    largestIdx = bootOrder > largestIdx ? bootOrder : largestIdx;
+  });
+
+  nics.forEach(nic => {
+    const bootOrder = get(nic, 'bootOrder', -1);
+    largestIdx = bootOrder > largestIdx ? bootOrder : largestIdx;
+  });
+
+  // assigned indexes start at two as the first index is assigned directly by the user
+  return largestIdx !== -1 ? largestIdx + 1 : BOOT_ORDER_SECOND;
+};
+
+export const sequentializeBootOrderIndexes = vm => {
+  const sortedBootableDevices = getBootableDevicesInOrder(vm);
+
+  let index;
+  for (index = 0; index <= sortedBootableDevices.length; index++) {
+    const device = sortedBootableDevices[index];
+    updateBootOrder(vm, device, index + 1);
+  }
+};
+
+export const getBootableDevices = (vm, vmDisks = [], vmNics = []) => {
+  const bootableDevices = [];
+  const disks = vm ? getDisks(vm) : vmDisks;
+  const nics = vm ? getInterfaces(vm) : vmNics;
+
+  disks.forEach(disk => {
+    const bootOrder = get(disk, 'bootOrder', null);
+    if (bootOrder) {
+      bootableDevices.push(disk);
+    }
+  });
+
+  nics.forEach(nic => {
+    const bootOrder = get(nic, 'bootOrder', null);
+    if (bootOrder) {
+      bootableDevices.push(nic);
+    }
+  });
+
+  return bootableDevices;
+};
+
+export const getBootableDevicesInOrder = vm => getBootableDevices(vm).sort((a, b) => a.bootOrder - b.bootOrder);
+
+export const isNic = device => !!get(device, 'bridge', null);
+
+export const isDisk = device => !!get(device, 'disk', null);
+
+export const updateBootOrder = (vm, device, bootOrder) => {
+  if (isNic(device)) {
+    getInterfaces(vm).forEach(nic => {
```

Comment:
why do we have to find that device again anyway? isn't nic and device the same exact object?

Reply:
This was me being stupid. Thanks for making me realize it. It allowed me to clean things up significantly.